File: day_04/solve.rb
FILE = ARGV[0] || "./test"

REQUIRED_FIELDS = {
  "byr" => /^19[2-9]\d|200[012]$/,
  "iyr" => /^201\d|2020$/,
  "eyr" => /^202\d|2030$/,
  "hgt" => /(^((59)|(6\d)|(7[0-6]))(?:in)$)|(^((1[5-8]\d)|(19[0-3]))(?:cm)$)/,
  "hcl" => /^#[0-9a-f]{6}$/,
  "ecl" => /^amb|blu|brn|gry|grn|hzl|oth$/,
  "pid" => /^\d{9}$/,
}

def run(validate_fields = false)
  curr_required_fields = []
  valid_passports = 0

  File.foreach(FILE) do |line|
    line.strip!
    if line.empty?
      valid_passports += REQUIRED_FIELDS.length == curr_required_fields.length ? 1 : 0
      curr_required_fields = []
      next
    end
    curr_required_fields += line.split(" ").map { |l| l.strip.split(":") }.reject { |f,v| f == 'cid' || (validate_fields && !REQUIRED_FIELDS[f].match?(v)) }.map { |f, v| f }
  end

  # last line not empty
  valid_passports += REQUIRED_FIELDS.length == curr_required_fields.length ? 1 : 0

  valid_passports
end

puts "First part (no value validation): #{run}"
puts "Second part (values validated): #{run(true)}"

File: 2020/day_04/solve.exs
valid_field? = fn
  "byr:" <> _, false ->
    true

  "iyr:" <> _, false ->
    true

  "eyr:" <> _, false ->
    true

  "hgt:" <> _, false ->
    true

  "hcl:" <> _, false ->
    true

  "ecl:" <> _, false ->
    true

  "pid:" <> _, false ->
    true

  "byr:" <> val, true ->
    Regex.match?(~r/^19[2-9]\d|200[012]$/, val)

  "iyr:" <> val, true ->
    Regex.match?(~r/^201\d|2020$/, val)

  "eyr:" <> val, true ->
    Regex.match?(~r/^202\d|2030$/, val)

  "hgt:" <> val, true ->
    Regex.match?(~r/(^((59)|(6\d)|(7[0-6]))(?:in)$)|(^((1[5-8]\d)|(19[0-3]))(?:cm)$)/, val)

  "hcl:" <> val, true ->
    Regex.match?(~r/^#[0-9a-f]{6}$/, val)

  "ecl:" <> val, true ->
    Regex.match?(~r/^amb|blu|brn|gry|grn|hzl|oth$/, val)

  "pid:" <> val, true ->
    Regex.match?(~r/^\d{9}$/, val)

  _, _ ->
    false
end

cond_inc = fn
  v, 7 -> v + 1
  v, _ -> v
end

solve = fn validate_values ->
  {r, f} =
    File.stream!("./input")
    |> Enum.map(&(String.trim(&1) |> String.split(" ", trim: true)))
    |> Enum.reduce({0, 0}, fn
      [], {c, f} ->
        {cond_inc.(c, f), 0}

      fields, {c, f} ->
        {c, f + Enum.count(fields, &valid_field?.(&1, validate_values))}
    end)

  cond_inc.(r, f)
end

IO.puts("First part: #{solve.(false)}")
IO.puts("Second part: #{solve.(true)}")
